### ISSUE TITLE
increase mouse sensitivity area for inlets & outlets

### ISSUE THREAD

**Issue by paulcarbone** [pure-data/pure-data#1926]
hello - 
I've been finding it hard to make connections on higher resolution displays, and had resorted to increasing the sizes of the inlets & outlets using the #define lines in canvas.h.  I've just moved to a 4k setup on a windows computer, and just don't want to keep re-compiling.  Is there any way to make this an option that can be set in preferences, or a way to have the connection mode get triggered (the 'o' icon) when the mouse is within a few pixels to the outside of the port, like on plugdata? 
thanks

**Comment by umlaeute** [comment]
Currently not.
You do know that you can "zoom in" to make everything larger, don't you?

**Comment by paulcarbone** [comment]
I do.  On a 4K screen, with the fonts cranked way down, it does help a little.  On lower res screens, the 2x zoom seems comically large.  I'd still like a little more meat on those inlets & outlets; guess I'll just keep recompiling my own.
thanks

**Comment by porres** [comment]
that was given up pretty quickly :) making in/outlets bigger doesn't look good to me, at least drawing them that big... not sure how to make it for sensitive for mouse nearby, but that seems more like it... why don't you rename and reopen this?

**Comment by umlaeute** [comment]
> I do. On a 4K screen, with the fonts cranked way down, it does help a little.

Good.

> On lower res screens, the 2x zoom seems comically large.

Well, I understood your issue to be about being able to patch on hidpi screens, so that shouldn't be problem for you then (and on low-res screens just don't zoom in...)

**Comment by paulcarbone** [comment]
Hey Alex - 
Yeah, the big in/outlets looks pretty goofy, but was the best solution I could come up with with my non-existent programming chops.  Increasing mouse sensitivity zone would be nicer, but I'm not sure how involved that would be.  I'll change the ask and see if anyone bites?

**Comment by umlaeute** [comment]
here's another "You do know":

You do know that you can use <kbd>Tab</kbd> to cycle through the outlets (and inlets, but doesn't work as well on some systems) when creating a connection?

https://vimeo.com/340437816#t=0h2m52s

**Comment by umlaeute** [comment]
@paulcarbone just to completely understand your issue: is your problem the width of the outlets, or the height of the outlets, or both?

extending the hotspot *width* should be no problem, but the *height* might be (as this interferes with other mouse interactions)

**Comment by paulcarbone** [comment]
It's both, but I'd say more so the height.  The I the in/outlets are only 3 pixels high in the default config, and it just seems to require more mousing precision than should be necessary.  
I did find that when I increased the sizes in canvas.h, I had to be careful with the width, because otherwise on smaller objects, the in/outlets would be too close to each other.
Here's a zoomed in image of what would be ideal - the in/outlets are currently I think 8px by 3px.  If the hotspot was say, even 3 pixels larger that could help a lot (as specially for differently abled folks). Is the current hotspot any larger than the in/outlet itself? It doesn't feel like it.
![image](https://user-images.githubusercontent.com/8267407/226497202-fea21499-22d8-4f8d-b17b-ec39dbde30e8.png)

**Comment by porres** [comment]
> but the height might be (as this interferes with other mouse interactions)

what do you mean?

**Comment by umlaeute** [comment]
> what do you mean?

If we extend the outlet height upwards, it will occlude parts of the object itself, thus users will have a harder time selecting the object contents (for editing).
If we extend it downwards, it will occlude existing connections, thus users will have a harder time selecting connections (for deletion).

I'm only talking about the "active area", not the visible outlet (which I think can stay as out is).

@paulcarbone just to make sure (because you keep mentioning them): I don't have any intention to do anything about the inlets. I think they already do everything that can be done

**Comment by umlaeute** [comment]
> If we extend it downwards, [...]

extending downwards so it extends beyond the object's active area actually requires a more serious rewrite.

**Comment by umlaeute** [comment]
could you try the `improvement/outlet-size` branch?

it does enlarge the outlet hotspot area considerably but mostly in the width.
the enlarged height *already* gets slightly in the way with my normal patching workflow, so I'm a bit cautious to enlarge it even more.

however, it could well be that together with the enlarged width the overall hidpi patching experience is now acceptable.

**Comment by porres** [comment]
> If we extend it downwards, it will occlude existing connections, thus users will have a harder time selecting connections (for deletion).

I noticed this happens in PlugData and it's very very annoying

**Comment by paulcarbone** [comment]
Cool, thanks! Gonna have to hop back into a mac to try this - I honestly don't even know where to start building it in Win10

**Comment by umlaeute** [comment]
a good start is [msw/README.txt](https://github.com/pure-data/pure-data/blob/master/msw/README.txt) (though it's a bit thin on the MSYS2/MinGW requirements).

you could also just grab the build artifacts from the [CI](https://git.iem.at/pd/pure-data/-/pipelines?ref=improvement%2Foutlet-size):
- [macOS](https://git.iem.at/pd/pure-data/-/jobs/artifacts/improvement%2Foutlet-size/download?job=macOS)
- [Windows/64bit](https://git.iem.at/pd/pure-data/-/jobs/artifacts/improvement%2Foutlet-size/download?job=w64)
- ...

**Comment by paulcarbone** [comment]
'build artifacts'.. didn't know that was a thing, thanks. 
It is easier to patch, I think even that extra pixel in height on the hotspot helped.  Having the active area the entire width caught me off guard for a sec, but I don't think it's a real stumbling block (and honestly I didn't notice the inlets always worked like that until now). 
You scaled everything up by a bit to as well, right?
thanks very much.  Do you think any of this will / can be made options in future builds?

**Comment by umlaeute** [comment]
Am 24. März 2023 00:56:06 MEZ schrieb paulcarbone ***@***.***>:
>You scaled everything up by a bit to as well, right?

Not sure what you mean, but most likely "no"

>thanks very much.  Do you think any of this will / can be made options in future builds? 


Certainly not an "option". Either it goes in, or not...



mfg.gss.owh
IOhannes

**Comment by paulcarbone** [comment]
re scaling - top one is what was downloaded from build artifacts, bottom one is downloaded .53-2.  If I change the display scaling, they both scale in proportion to each other, along with all other apps.
![image](https://user-images.githubusercontent.com/8267407/227556622-e4c086fb-5461-43ac-a7c7-fa19e9731652.png)

**Comment by umlaeute** [comment]
one thing that changed was the default font size for the PdConsole and the menus (#1794)

that wouldn't explain the size difference of the object in the patch though.
i note however, that you get an error in the Pd-console for the Pd on top, maybe this is related?

**Comment by paulcarbone** [comment]
hmm.. they're both at the same font size, both console & patch window.  The error was unrelated - externals I haven't installed yet on windows. 
Here they are both at my usual 125% setting.
![image](https://user-images.githubusercontent.com/8267407/227564064-895d980d-fabc-489c-ba95-24f8e1af954f.png)

Here they are at 100%
![image](https://user-images.githubusercontent.com/8267407/227564420-49208056-81f0-45aa-a14b-3a4cd73a63ac.png)

I don't believe there's any way in windows to set per application scaling.

I'm going to try this on my (somewhat old) mac this weekend.  I followed the windows build instructions, which 'succeeded' but the binary wouldn't launch.

**Comment by umlaeute** [comment]
> I followed the windows build instructions, which 'succeeded' but the binary wouldn't launch.

i typically just run `make app`, which gives me a useable app (modulo some `libwinpthread-1.dll` issues on amd64; but I *think* this a cross-compiling issue)

**Comment by umlaeute** [comment]
> hmm.. they're both at the same font size, both console & patch window. 

the font-size of the patch is stored inside the patch (so the default font-size for patches only affects newly created patches).
the font-size for the PdConsole is stored in some global settings, as soon as you change them. (so the default font-size for the PdConsole only affects systems that never played with the font-size settings for the PdConsole).

in any case: i only mentioned them, because it is the only thing i think has changed that *might* effect the size rendering. maybe @Lucarda can shed some light on that.

in any case: the scaling is certainly not related to "increase mouse sensitivity area for inlets & outlets" so we probably should return to the topic :-)

**Comment by Lucarda** [comment]
hi @paulcarbone,


based on your pics above I think your "downloaded .53-2" is abnormal.

here is the "build artifact" which has the normal spacing counting from "menu help" to "audio off". see the green brush.


![size-a](https://user-images.githubusercontent.com/25239522/227669508-d2847562-90a6-42ed-a048-ae5c0c23c002.png)


and here is your "downloaded .53-2" with the abnormal spacing. see red brush.


![size-b](https://user-images.githubusercontent.com/25239522/227669533-87e042a3-b070-4e96-8780-6811ce653393.png)


----------------------

- are you sure you did not made any changes in the tcl files of your "downloaded .53-2"? 

  - if so can you uninstall and then reinstall the app?
  
  
---------------------

### some other stuff:

theres a difference in both apps you are testing but this seems unrelated to your issue:

"downloaded .53-2" runs on top of `wish/Tk 8.6.10` and the "build artifact" on top of `wish/Tk 8.5.10`

### about your build

>  I followed the windows build instructions, which 'succeeded' but the binary wouldn't launch.

it might be the case that you are trying to lunch:

`XX:\pure-data-improvement-outlet-size\bin\pd.exe` (this one is not supposed to lunch)

instead lunch:

`XX:\pure-data-improvement-outlet-size\pd-0.53.1\bin\pd.exe`

**Comment by Lucarda** [comment]
@paulcarbone if something unrelated to "increase mouse sensitivity area for inlets & outlets" arise please open another issue.

:)

**Comment by paulcarbone** [comment]
@umlaeute - I'm going to close out this issue as it relates to patching - I believe the larger hotspot does make it easier on high DPI displays, thank you! 
@Lucarda - I can open up another issue about this if you think it's helpful for troubleshooting, but I did re-download 53.2 and several older versions, and everything I tried back to .50 was smaller and with the abnormal spacing you called out like in the screenshot above. Not an issue per se for me, but an oddity & curiosity.

**Comment by umlaeute** [comment]
our usual policy is to close bugs/improvements/feature-requests once they are either
- accepted into `master`
- rejected
- obsolete

none is the case here, so i'm re-opening this (it will be automatically closed once the fix in PR #1929 is merged into `master`)

**Comment by umlaeute** [comment]
PS: i just noticed what could have caused side differences between the official releases and the CI-builds.

- the CI uses the TclTk that is included in [msw/pdprototype.tgz](https://github.com/pure-data/pure-data/blob/f2b7874669bb7c7cbe03afeb6556068c1c2ce6c7/msw/pdprototype.tgz), which happens to be Tcl/Tk-8.5.10
- the official Pd-0.53-1/Win32 package includes Tcl/Tk-8.5.19
- the official Pd-0.53-1/Win64 package includes Tcl/Tk-8.6.10

see also:
https://github.com/pure-data/pure-data/issues/1725#issuecomment-1230153972
